File: internal/application/visitorService.go
package application

import (
	"fmt"
	"time"
	"visitor-management/internal/domain"
	"visitor-management/internal/ports"

	"github.com/google/uuid"
)

type VisitorService struct {
	Repo     ports.VisitorRepository
	UserRepo ports.UserRepository
}

func NewVisitorService(repo ports.VisitorRepository, UserRepo ports.UserRepository) *VisitorService {
	return &VisitorService{Repo: repo, UserRepo: UserRepo}
}

func (s *VisitorService) CreateVisitor(v *domain.Visitor, userID string, role string) (*domain.Visitor, error) {

	v.ID = uuid.New().String() // visitor_id
	v.CreatedAt = time.Now().Unix()
	v.AddedByRole = role

	if role == "owner" {
		v.OwnerID = userID

		user, err := s.UserRepo.GetUserById(userID)
		if err != nil {
			return nil, fmt.Errorf("failed to fetch user: %w", err)
		}

		v.OwnerEmail = user.Email

	} else {

		owner, err := s.UserRepo.GetOwnerByTowerAndFlat(v.Tower, v.FlatNo)
		if err != nil {
			return nil, fmt.Errorf("failed to find owner: %w", err)
		}

		v.OwnerID = owner.ID
		v.OwnerEmail = owner.Email
	}

	if role == "admin" || role == "owner" {
		v.Status = "approved"
	} else {
		v.Status = "pending"
	}

	// Save to DynamoDB
	err := s.Repo.Create(v)
	if err != nil {
		return nil, err
	}

	// Return the full visitor object including ID
	return v, nil
}

// func (s *VisitorService) DeleteVisitor(id int) error {
// 	return s.Repo.Delete(id)
// }

func (s *VisitorService) GetAllVisitors() ([]domain.Visitor, error) {
	visitors, err := s.Repo.GetAllVisitors()
	if err != nil {
		fmt.Printf("Service error in GetAllVisitors: %v\n", err)
	}
	return visitors, err
}

func (s *VisitorService) GetVisitorsByOwner(ownerID string) ([]domain.Visitor, error) {
	visitors, err := s.Repo.GetVisitorsByOwner(ownerID)
	if err != nil {
		fmt.Printf("Service error in GetVisitorsByOwner (ownerID=%s): %v\n", ownerID, err)
	}
	return visitors, err
}
func (s *VisitorService) GetVisitorCount() (int, error) {
	return s.Repo.CountVisitors()
}
func (s *VisitorService) UpdateVisitorStatus(visitorID, ownerID, status string) error {

	return s.Repo.UpdateVisitorStatus(visitorID, ownerID, status)
}

File: src/services/visitor_service.py
from typing import List
import uuid
import time
from fastapi import HTTPException, status

from src.models.visitor import Visitor, VisitorStatus
from src.repository.visitor_repository import VisitorRepository
from src.repository.user_repository import UserRepository
from src.dto.visitor_dto import CreateVisitorRequest, VisitorResponse
from src.utils.mapper import to_visitor_response
from src.constants.role_enum import UserRole
from src.errors import error


class VisitorService:

    def __init__(self, visitor_repo: VisitorRepository, user_repo: UserRepository):
        self.visitor_repo = visitor_repo
        self.user_repo = user_repo

    def create_visitor(
        self,
        req: CreateVisitorRequest,
        user_id: str,
        role: UserRole,
    ) -> VisitorResponse:

        try:
            if role == UserRole.OWNER:
                owner_id = user_id
                user = self.user_repo.get_user_by_id(user_id)
                owner_email = user.Email
            else:
                owner = self.user_repo.get_owner_by_tower_and_flat(
                    req.tower,
                    req.flat_no,
                )
                owner_id = owner.ID
                owner_email = owner.Email
        except error.UserNotFoundError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Owner not found",
            )

        if role in (UserRole.ADMIN, UserRole.OWNER):
            visitor_status = VisitorStatus.APPROVED
        else:
            visitor_status = VisitorStatus.PENDING

        visitor = Visitor(
            ID=str(uuid.uuid4()),
            Name=req.name,
            Email=req.email,
            Tower=req.tower,
            FlatNo=req.flat_no,
            AddedByRole=role.value,
            CreatedAt=int(time.time()),
            Status=visitor_status,
            OwnerID=owner_id,
            OwnerEmail=owner_email,
        )

        try:
            self.visitor_repo.create(visitor)
        except error.RepositoryError as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create visitor",
            ) from e

        return to_visitor_response(visitor)

    def get_all_visitors(self) -> List[VisitorResponse]:
        visitors = self.visitor_repo.get_all_visitors()
        return [to_visitor_response(v) for v in visitors]

    def get_visitors_by_owner(self, owner_id: str) -> List[VisitorResponse]:
        visitors = self.visitor_repo.get_visitors_by_owner(owner_id)
        return [to_visitor_response(v) for v in visitors]

    def get_visitors_count(self) -> int:
        return self.visitor_repo.count_visitors()

    def get_count_visitors_by_owner(self,user_id: str) -> int:
        return self.visitor_repo.count_visitors_by_owner(user_id)

    def update_visitor_status(
        self,
        visitor_id: str,
        owner_id: str,
        status_value: VisitorStatus,
    ) -> None:
        try:
            self.visitor_repo.update_visitor_status(visitor_id, owner_id, status_value)
        except error.NotFoundError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Visitor not found",
            )
        except error.RepositoryError as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to update visitor status",
            ) from e
